torch/distributed/_tensor/experimental/attention.py

```diff
@@ -634,27 +662,198 @@ def backward_hook(grad: torch.Tensor) -> None:
+                setattr_(fn_caller_module, obj_name, obj, wrapper_func)
+
+
+def enable_context_parallel(
+    seq_dim: int,
+    callers: List[nn.Module],
+    device_mesh: DeviceMesh,
+) -> None:
+    """
+    This is an experimental API to enable context parallelism for
+    ``torch.nn.functional.scaled_dot_product_attention``. This API assumes
+    that the q, k, v are already sharded on the ``seq_dim`` dimension and
+    will install hook to convert the q, k, v to the DTensors.
+
+    This API will change ``scaled_dot_product_attention`` in ``torch.nn.functional``
+    (short as ``F``) to a wrapped function. So any subsequent call to
+    ``F.scaled_dot_product_attention`` will be redirected to the wrapped function.
+
+    Note that it is important to include all the modules that call SDPA in the
+    ``callers`` list. This can avoid the incorrect wrapping if the model code uses
+    ``from ... import ..." to import SDPA.
+
+    Args:
+        seq_dim (int): the sequence dimension for q, k, v.
+        callers (List[nn.Module]): the nn.Modules that call ``scaled_dot_product_attention``.
+        device_mesh (:class:`DeviceMesh`, optional): the device mesh for context
+            parallelism.
+    """
+
+    def attention_input_fn(
+        device_mesh: DeviceMesh, *args: Tuple[Any, ...], **kwargs: Dict[str, Any]
+    ) -> Tuple[Tuple[Any, ...], Dict[str, Any]]:
+        placement = [Shard(seq_dim)]
+        all_args = []
+
+        for arg in itertools.chain(args, kwargs.values()):
+            if isinstance(arg, torch.Tensor) and not isinstance(arg, DTensor):
+                arg = DTensor.from_local(arg, device_mesh, placement, run_check=False)
+
+            all_args.append(arg)
+
+        new_args = tuple(all_args[0 : len(args)])
+        new_kwargs = dict(zip(kwargs.keys(), all_args[len(args) :]))
+        return new_args, new_kwargs
+
+    def attention_output_fn(device_mesh: DeviceMesh, outputs: Any) -> Any:
+        new_outputs = []
+        for output in [outputs] if isinstance(outputs, torch.Tensor) else outputs:
+            output = output.to_local() if isinstance(output, DTensor) else output
+            new_outputs.append(output)
+
+        if isinstance(outputs, torch.Tensor):
+            return new_outputs[0]
+
+        return tuple(new_outputs)
+
+    old_handlers = DTensor._op_dispatcher._custom_op_handlers
+    DTensor._op_dispatcher._custom_op_handlers = {**old_handlers, **customized_ops}
+    _distribute_function(
+        F.scaled_dot_product_attention,
+        F,
+        callers,
+        device_mesh,
+        attention_input_fn,
+        attention_output_fn,
+    )
+
+
+def _get_sequence_shard(
+    buffer: torch.Tensor, mesh: DeviceMesh, seq_dim: int
+) -> torch.Tensor:
+    return buffer.chunk(mesh.size(), dim=seq_dim)[mesh.get_local_rank()]
+
+
+@contextlib.contextmanager
+@torch.no_grad()
+def context_parallel_buffers(
+    mesh: Optional[DeviceMesh],
+    buffers: List[torch.Tensor],
```

Comment:
Could you add documentation on this API, specifically I am confused about what are the meanings of `buffers`, and `restore_funcs`?

Reply:
I merged and remove the APIs. Now we only have one public API, `context_parallel()`.